### PR TITLE
refactor(query): refactor hash join spill

### PR DESCRIPTION
**Pull request by Dousir9**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

### Code readability and maintainability
1. New step transition mechanism: this PR divides hash join probe/build step into three types: `SyncStep`, `AsyncStep`, and `Finish`, ensuring that step transition occurs in the event and this PR redesigns step transition to make the step transition logic clearer and improve code readability and maintainability.
2. Unify `ProbeSpillHandler` and `BuildSpillHandler` to `HashJoinSpiller` to improve code maintainability.
3. Unify the spill logic of cross join and other types.
4. Introduce `HashTableType` to reduce code.
### Bug Fixed
1. Fix a corner case of left-related join when spill happens.
2. The left-anti join should perform probe if there is no partition on the build side.
3. Fix out of range for slice panic when spill happens.
### Tests
- [x] Add TPC-H spill test
- [x] Add TPC-DS spill test
### Performance (Databend Cloud XSMALL)
TPC-H SF100 Q18: **164s → 137s**

Test script:
```sql
set join_spilling_buffer_threshold_per_proc_mb = 512;
set join_spilling_bytes_threshold_per_proc = 1;
set join_spilling_memory_ratio = 1;
select
    c_name,
    c_custkey,
    o_orderkey,
    o_orderdate,
    o_totalprice,
    sum(l_quantity)
from
    customer,
    orders,
    lineitem
where
        o_orderkey in (
        select
            l_orderkey
        from
            lineitem
        group by
            l_orderkey having
                sum(l_quantity) > 300
    )
  and c_custkey = o_custkey
  and o_orderkey = l_orderkey
group by
    c_name,
    c_custkey,
    o_orderkey,
    o_orderdate,
    o_totalprice
order by
    o_totalprice desc,
    o_orderdate;
```

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [x] Performance Improvement
- [x] Other (please describe): code maintainability

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15746)

### CODE REVIEW

File: src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_build_state.rs [hunk 1]
```diff
@@ -285,20 +282,18 @@ impl HashJoinBuildState {
...
+        self.next_round_counter.fetch_add(1, Ordering::Relaxed);
     }
 
-    /// Detach to state: `row_space_builders`,
+    /// Detach to state: `collect_counter`,
     /// create finalize task and initialize the hash table.
-    pub(crate) fn row_space_build_done(&self) -> Result<()> {
-        let old_count = self.row_space_builders.fetch_sub(1, Ordering::Relaxed);
+    pub(crate) fn collect_done(&self) -> Result<()> {
+        let old_count = self.collect_counter.fetch_sub(1, Ordering::Relaxed);
```
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_build_state.rs:296]: Is the Relaxed level sufficient?

**Review by zhang2014 — APPROVED** [review]
LGTM

**Review by xudong963 — APPROVED** [review]
